ethdb/privateapi/ethbackend.go

```diff
@@ -271,18 +282,31 @@ func (s *EthBackendServer) EngineExecutePayloadV1(ctx context.Context, req *type
 
 // EngineGetPayloadV1, retrieves previously assembled payload (Validators only)
 func (s *EthBackendServer) EngineGetPayloadV1(ctx context.Context, req *remote.EngineGetPayloadRequest) (*types2.ExecutionPayload, error) {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	// Wait some time in case validition process has not started
+	timer := time.NewTimer(100 * time.Millisecond)
+	select {
+	case <-timer.C:
+	case <-ctx.Done():
+		return nil, fmt.Errorf("execution layer interrupted")
+	}
+
+	if !s.proposing {
+		return nil, fmt.Errorf("execution layer not running as a proposer. enable --proposer flag on startup")
+	}
 
 	if s.config.TerminalTotalDifficulty == nil {
 		return nil, fmt.Errorf("not a proof-of-stake chain")
 	}
-
+	s.mu.Lock()
+	defer s.mu.Unlock()
 	payload, ok := s.pendingPayloads[req.PayloadId]
-	if ok {
-		return &payload, nil
+	if !ok || payload.BlockNumber == 0 {
```

Comment:
What if it's a known payload, but its block assembly hasn't finished yet?

Reply:
if its block assembly has not finished yet, then blockNumber will be equal to 0 hence the second condition

Reply:
But then we shouldn't return `fmt.Errorf("unknown payload")`, we should wait for the assembly to finish.